File: heap/heap.py
# Implementation of a max heap
# Arrays are used to implement this heap
# Any node is greater than either of the children
# To get the children of a node at index n, look at index 2n and 2n+1
#
#
# Author: Prashant Malani <p.malani@gmail.com>



class Heap:
  def __init__(self, element_list):
    """Initialize the heap array with a few value. Doesn't have the heap
    property.
    """
    self.array = element_list[:]
    self.ConstructHeap()

  def MaxHeapify(self, i):
    """ Index i is where you want to start the heapification.
    """
    l = 2*i + 1
    r = 2*i + 2
    largest = i
    if l <= len(self.array) - 1:
      if self.array[l] > self.array[i]:
        largest = l
    if r <= len(self.array) - 1:
      if self.array[r] > self.array[largest]:
        largest = r
    if largest is not i:
      swap = self.array[i]
      self.array[i] = self.array[largest]
      self.array[largest] = swap
      self.MaxHeapify(largest)

  def ConstructHeap(self):
    """ Convert the array into a heap, basically call max_heapify on all the  children which have
    at most 2 children (max_heapify needs both sub children to have heap property
    """
    end_val = len(self.array) - 1
    end_val = (end_val - 1) / 2
    list_of_indices = range(end_val + 1)
    list_of_indices.reverse()
    for i in list_of_indices:
      self.MaxHeapify(i)

File: heap/heap.c
/*
 * HeapSort C Implementation
 *
 * Author : Prashant Malani <p.malani@gmail.com>
 * Date : 02/03/2014
 *
 * Heap is implemented as an array
 */
#include <stdio.h>
#include <stdlib.h>

/* Helper routine */
void swap(int *a, int *b)
{
	int temp;
	temp = *a;
	*a = *b;
	*b = temp;
}

/*
 * Heap implementation routine used in both heap construction
 * and heapsort.
 */
void heapify(int index, int a[], int len)
{
	int l,r,high;
	l = 2 * index + 1;
	r = 2 * index + 2;
	high = index;
	if (l < len && a[l] > a[high])
		high = l;
	if (r < len && a[r] > a[high])
		high = r;
	if (high != index) {
		swap(&a[high], &a[index]);
		heapify(high, a, len);
	}
	return;
}

void construct_heap(int arr[], int len)
{
	int i;
	for (i = (len - 1) / 2; i >= 0; i--)
		heapify(i, arr, len);
}

void do_heapsort(int a[], int len)
{
	int i = len - 1;
	while (i > 0) {
		/*
		 * Swap first element with last, move to the top, reduce length,
		 * the call heapify on new length array.
		 */
		swap(&a[0], &a[i]);
		i--;
		heapify(0, a, i);
	}
}

int main()
{
	int arr[] = {34, -3, 1983, -12342, 883, 74 };
	int len = sizeof(arr) / sizeof(arr[0]);
	int i;

	construct_heap(arr, len);
	do_heapsort(arr, len);

	for (i = 0; i < len; i++)
		printf("%d ", arr[i]);
	printf("\n");
	return 0;
}
